File: Kenesbek/hw6/hack_assembler.py
comp_table = {
    "0": "0101010", "1": "0111111", "-1": "0111010",
    "D": "0001100", "A": "0110000", "M": "1110000",
    "!D": "0001101", "!A": "0110001", "!M": "1110001",
    "-D": "0001111", "-A": "0110011", "-M": "1110011",
    "D+1": "0011111", "A+1": "0110111", "M+1": "1110111",
    "D-1": "0001110", "A-1": "0110010", "M-1": "1110010",
    "D+A": "0000010", "D+M": "1000010", "D-A": "0010011",
    "D-M": "1010011", "A-D": "0000111", "M-D": "1000111",
    "D&A": "0000000", "D&M": "1000000", "D|A": "0010101", "D|M": "1010101"
}

dest_table = {
    "": "000", "M": "001", "D": "010", "MD": "011",
    "A": "100", "AM": "101", "AD": "110", "AMD": "111"
}

jump_table = {
    "": "000", "JGT": "001", "JEQ": "010", "JGE": "011",
    "JLT": "100", "JNE": "101", "JLE": "110", "JMP": "111"
}

# Initialize symbol table with predefined symbols
symbol_table = {
    "SP": 0, "LCL": 1, "ARG": 2, "THIS": 3, "THAT": 4,
    "SCREEN": 16384, "KBD": 24576
}
symbol_table.update({f"R{i}": i for i in range(16)})  # R0 to R15

def clean_line(line):
    """Remove comments and whitespace."""
    line = line.split("//")[0].strip()  # Remove comments and trailing whitespace
    return line

def translate_a_instruction(instruction):
    """Translate A-instruction (e.g., @2) to binary."""
    value = int(instruction[1:])  # Get number after "@"
    return "0" + format(value, "015b")  # Convert to 15-bit binary prefixed by 0

def parse_c_instruction(instruction):
    """Separate comp, dest, and jump parts from a C-instruction."""
    if "=" in instruction:
        dest, remainder = instruction.split("=")
    else:
        dest, remainder = "", instruction

    if ";" in remainder:
        comp, jump = remainder.split(";")
    else:
        comp, jump = remainder, ""

    return dest, comp, jump

def translate_c_instruction(instruction):
    """Translate C-instruction to binary using lookup tables."""
    dest, comp, jump = parse_c_instruction(instruction)
    return "111" + comp_table[comp] + dest_table[dest] + jump_table[jump]

def first_pass(lines):
    """Process labels and add them to the symbol table."""
    line_number = 0
    for line in lines:
        if line.startswith("(") and line.endswith(")"):
            label = line[1:-1]
            symbol_table[label] = line_number  # Label points to this line
        else:
            line_number += 1

def second_pass(lines):
    """Replace symbols (variables) with addresses in the symbol table."""
    variable_address = 16
    output_lines = []
    for line in lines:
        # Skip labels
        if line.startswith("(") and line.endswith(")"):
            continue

        if line.startswith("@"):
            symbol = line[1:]
            if not symbol.isdigit():
                if symbol not in symbol_table:
                    symbol_table[symbol] = variable_address  # Assign new variable address
                    variable_address += 1
                address = symbol_table[symbol]
                output_lines.append(f"@{address}")
            else:
                output_lines.append(line)
        else:
            output_lines.append(line)
    return output_lines

def assemble(input_file, output_file):
    with open(input_file, "r") as file:
        lines = [clean_line(line) for line in file if clean_line(line)]

    first_pass(lines)

    processed_lines = second_pass(lines)

    with open(output_file, "w") as file:
        for line in processed_lines:
            if line.startswith("@"):
                binary_code = translate_a_instruction(line)
            else:
                binary_code = translate_c_instruction(line)
            file.write(binary_code + "\n")

File: Borsos/hw6/HackAssembler.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

#define MAX_LINE_LENGTH 256
#define SYMBOL_TABLE_SIZE 100

// Instruction Types
typedef enum { A_INSTRUCTION, C_INSTRUCTION, L_INSTRUCTION } InstructionType;

// Structure for a symbol in the symbol table
typedef struct {
    char symbol[MAX_LINE_LENGTH];
    int address;
} Symbol;

// Symbol table array and current index
Symbol symbolTable[SYMBOL_TABLE_SIZE];
int symbolCount = 0;
int variableAddress = 16;  // Start variables at address 16

// Function to remove whitespace and comments from a line
void clean_line(char *line) {
    char *comment = strchr(line, '/');
    if (comment) *comment = '\0';  // Remove comments
    // Trim whitespace
    char *end = line + strlen(line) - 1;
    while (end > line && isspace((unsigned char)*end)) end--;
    end[1] = '\0';
}

// Add a symbol to the symbol table
void add_symbol(const char *symbol, int address) {
    strcpy(symbolTable[symbolCount].symbol, symbol);
    symbolTable[symbolCount].address = address;
    symbolCount++;
}

// Get the address of a symbol from the table, or -1 if not found
int get_symbol_address(const char *symbol) {
    for (int i = 0; i < symbolCount; i++) {
        if (strcmp(symbolTable[i].symbol, symbol) == 0) {
            return symbolTable[i].address;
        }
    }
    return -1; // Symbol not found
}

// Initialize predefined symbols in Hack assembly
void initialize_symbol_table() {
    for (int i = 0; i <= 15; i++) {
        char reg[4];
        sprintf(reg, "R%d", i);
        add_symbol(reg, i);
    }
    add_symbol("SCREEN", 16384);
    add_symbol("KBD", 24576);
    add_symbol("SP", 0);
    add_symbol("LCL", 1);
    add_symbol("ARG", 2);
    add_symbol("THIS", 3);
    add_symbol("THAT", 4);
}

void print_symbol_table() {
    for (int i = 0; i < symbolCount; i++) {
        printf("%s %d\n", symbolTable[i].symbol, symbolTable[i].address);
    }
}

// Determine the type of instruction (A, C, or L)
InstructionType get_instruction_type(char *line) {
    if (line[0] == '@') return A_INSTRUCTION;
    if (line[0] == '(' && line[strlen(line) - 1] == ')') return L_INSTRUCTION;
    return C_INSTRUCTION;
}

// Parse the symbol from an A-instruction or L-instruction
void parse_symbol(char *line, char *symbol) {
    if (line[0] == '@') {  // A-instruction
        strcpy(symbol, line + 1);
    } else {               // L-instruction
        strncpy(symbol, line + 1, strlen(line) - 2);
        symbol[strlen(line) - 2] = '\0';
    }
}

// Translate an A-instruction to binary
void translate_a_instruction(char *line, FILE *output) {
    char symbol[MAX_LINE_LENGTH];
    parse_symbol(line, symbol);
    int address;

    // Check if symbol is a number or a variable
    if (isdigit(symbol[0])) {
        address = atoi(symbol);
    } else {
        int symbolAddress = get_symbol_address(symbol);
        if (symbolAddress == -1) {  // Not in symbol table
            address = variableAddress;
            add_symbol(symbol, variableAddress);
            variableAddress++;
        } else {
            address = symbolAddress;
        }
    }

    // Convert address to 16-bit binary
    char binary[17] = "0000000000000000";
    for (int i = 15; i >= 0; i--) {
        binary[i] = (address % 2) ? '1' : '0';
        address /= 2;
    }
    fprintf(output, "%s\n", binary);
}

// Function to get binary code for dest
const char* get_dest_code(const char *dest) {
    if (strcmp(dest, "") == 0)          return "000";
    else if (strcmp(dest, "M") == 0)    return "001";
    else if (strcmp(dest, "D") == 0)    return "010";
    else if (strcmp(dest, "MD") == 0)   return "011";
    else if (strcmp(dest, "A") == 0)    return "100";
    else if (strcmp(dest, "AM") == 0)   return "101";
    else if (strcmp(dest, "AD") == 0)   return "110";
    else if (strcmp(dest, "AMD") == 0)  return "111";
    return "000"; // default if no dest part
}

// Function to get binary code for jump
const char* get_jump_code(const char *jump) {
    if (strcmp(jump, "") == 0)         return "000";
    else if (strcmp(jump, "JGT") == 0) return "001";
    else if (strcmp(jump, "JEQ") == 0) return "010";
    else if (strcmp(jump, "JGE") == 0) return "011";
    else if (strcmp(jump, "JLT") == 0) return "100";
    else if (strcmp(jump, "JNE") == 0) return "101";
    else if (strcmp(jump, "JLE") == 0) return "110";
    else if (strcmp(jump, "JMP") == 0) return "111";
    return "000"; // default if no jump part
}

// Function to get binary code for comp
const char* get_comp_code(const char *comp) {
    if (strcmp(comp, "0") == 0)              return "0101010";
    else if (strcmp(comp, "1") == 0)         return "0111111";
    else if (strcmp(comp, "-1") == 0)        return "0111010";
    else if (strcmp(comp, "D") == 0)         return "0001100";
    else if (strcmp(comp, "A") == 0)         return "0110000";
    else if (strcmp(comp, "!D") == 0)        return "0001101";
    else if (strcmp(comp, "!A") == 0)        return "0110001";
    else if (strcmp(comp, "-D") == 0)        return "0001111";
    else if (strcmp(comp, "-A") == 0)        return "0110011";
    else if (strcmp(comp, "D+1") == 0)       return "0011111";
    else if (strcmp(comp, "A+1") == 0)       return "0110111";
    else if (strcmp(comp, "D-1") == 0)       return "0001110";
    else if (strcmp(comp, "A-1") == 0)       return "0110010";
    else if (strcmp(comp, "D+A") == 0)       return "0000010";
    else if (strcmp(comp, "D-A") == 0)       return "0010011";
    else if (strcmp(comp, "A-D") == 0)       return "0000111";
    else if (strcmp(comp, "D&A") == 0)       return "0000000";
    else if (strcmp(comp, "D|A") == 0)       return "0010101";
    else if (strcmp(comp, "M") == 0)         return "1110000";
    else if (strcmp(comp, "!M") == 0)        return "1110001";
    else if (strcmp(comp, "-M") == 0)        return "1110011";
    else if (strcmp(comp, "M+1") == 0)       return "1110111";
    else if (strcmp(comp, "M-1") == 0)       return "1110010";
    else if (strcmp(comp, "D+M") == 0)       return "1000010";
    else if (strcmp(comp, "D-M") == 0)       return "1010011";
    else if (strcmp(comp, "M-D") == 0)       return "1000111";
    else if (strcmp(comp, "D&M") == 0)       return "1000000";
    else if (strcmp(comp, "D|M") == 0)       return "1010101";
    return "0101010"; // default if comp part is not recognized
}

void translate_c_instruction(char *line, FILE *output) {
    char dest[4] = "", comp[8] = "", jump[4] = "";

    // Check for destination (dest=comp;jump)
    char *equal_sign = strchr(line, '=');
    char *semicolon = strchr(line, ';');

    if (equal_sign) {
        // Parse destination and computation parts
        strncpy(dest, line, equal_sign - line);
        dest[equal_sign - line] = '\0';
        if (semicolon) {
            strncpy(comp, equal_sign + 1, semicolon - equal_sign - 1);
            comp[semicolon - equal_sign - 1] = '\0';
            strcpy(jump, semicolon + 1);
        } else {
            strcpy(comp, equal_sign + 1);
        }
    } else {
        // No destination, just computation and possibly jump
        if (semicolon) {
            strncpy(comp, line, semicolon - line);
            comp[semicolon - line] = '\0';
            strcpy(jump, semicolon + 1);
        } else {
            strcpy(comp, line);
        }
    }

    // Get binary codes for each part
    const char *comp_code = get_comp_code(comp);
    const char *dest_code = get_dest_code(dest);
    const char *jump_code = get_jump_code(jump);

    // Construct the full binary C-instruction
    fprintf(output, "111%s%s%s\n", comp_code, dest_code, jump_code);
}

int main() {
    char* input_files[7] = {"./Add/Add.asm", "./Max/Max.asm", "./Max/MaxL.asm", "./Rect/Rect.Asm", "./Rect/RectL.Asm", "./Pong/Pong.asm", "./Pong/PongL.asm"};
    char* output_files[7] = {"./Add/Add.hack", "./Max/Max.hack", "./Max/MaxL.hack", "./Rect/Rect.hack", "./Rect/RectL.hack", "./Pong/Pong.hack", "./Pong/PongL.hack"};

    for( int i=0; i<5; i++ ) 
    {
        FILE *input = fopen(input_files[i], "r");
        FILE *output = fopen(output_files[i], "w");

        if (!input || !output) {
            fprintf(stderr, "Error opening files.\n");
            return 1;
        }

        initialize_symbol_table();

        // First pass: process labels (L-instructions)
        int romAddress = 0;
        char line[MAX_LINE_LENGTH];
        while (fgets(line, sizeof(line), input)) {
            clean_line(line);
            if (line[0] == '\0' || line[1] == '\0') continue;
            InstructionType type = get_instruction_type(line);
            if (type == L_INSTRUCTION) {
                char symbol[MAX_LINE_LENGTH];
                parse_symbol(line, symbol);
                add_symbol(symbol, romAddress);
            } else {
                romAddress++;
            }
        }

        // Rewind file for second pass
        rewind(input);

        // Second pass: translate instructions
        while (fgets(line, sizeof(line), input)) {
            clean_line(line);
            if (line[0] == '\0' || line[1] == '\0') continue;

            InstructionType type = get_instruction_type(line);
            if (type == A_INSTRUCTION) {
                translate_a_instruction(line, output);
            } else if (type == C_INSTRUCTION) {
                translate_c_instruction(line, output);
            }
        }

        fclose(input);
        fclose(output);
    }
    return 0;
}
